src/metabase/query_processor/middleware/add_query_throttle.clj

```diff
@@ -0,0 +1,48 @@
+(ns metabase.query-processor.middleware.add-query-throttle
+  "Middleware that constraints the number of concurrent queries, rejects queries by throwing an exception and
+  returning a 503 when we exceed our capacity"
+  (:require [metabase.config :as config]
+            [puppetlabs.i18n.core :refer [tru]])
+  (:import [java.util.concurrent Semaphore TimeUnit]))
+
+(defn- calculate-max-queries-from-max-threads []
+  (let [max-threads (or (config/config-int :mb-jetty-maxthreads) 50)]
+    (int (Math/ceil (/ max-threads 2)))))
+
+(defn- ^Semaphore create-query-semaphore []
+  (let [total-concurrent-queries (or (config/config-int :mb-max-concurrent-queries)
+                                     (calculate-max-queries-from-max-threads))]
+    (Semaphore. total-concurrent-queries true)))
+
+(def ^Semaphore ^:private query-semaphore (create-query-semaphore))
+
+(defn- throw-503-unavailable
+  []
+  (throw (ex-info (tru "Max concurrent query limit reached")
+           {:status 503})))
```

Comment:
I think it's supposed to be `status-code`
But the QP middleware that catches exceptions automatically is going to catch this so we need to put some logic there too to make sure the exceptions come back the way we want

Reply:
or maybe it's already working?